fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/util/JDBCQueryBuilder.java

```diff
@@ -1031,9 +1055,19 @@ protected SqlQueryData processInclusionCriteria(QueryParameter queryParm) throws
                 whereClauseSegment.append(PARAMETER_TABLE_ALIAS).append(DOT).append(TOKEN_VALUE).append(operator)
                         .append(BIND_VAR);
                 if (resourceTypeName != null) {
-                    int codeSystemIdForResourceType = getCodeSystemId(resourceTypeName);
+                    Integer codeSystemIdForResourceType = getCodeSystemId(resourceTypeName);
                     whereClauseSegment.append(AND).append(PARAMETER_TABLE_ALIAS).append(DOT)
-                        .append(CODE_SYSTEM_ID).append(EQ).append(codeSystemIdForResourceType);
+                        .append(CODE_SYSTEM_ID).append(EQ).append(nullCheck(codeSystemIdForResourceType));
+
+                    // For #1929 add common_token_value_id = <n> for better query optimization
+                    Long commonTokenValueId = getCommonTokenValueId(resourceTypeName, currentParmValue);
+                    if (commonTokenValueId != null) {
+                        // #1929 improves cardinality estimation
+                        // resulting in far better execution plans for many search queries. Because COMMON_TOKEN_VALUE_ID
+                        // is the primary key for the common_token_values table, we don't need the CODE_SYSTEM_ID = ? predicate.
```

Comment:
Your comment says we don't need the CODE_SYSTEM_ID = ? predicate. Should lines 1058-1060 be moved to an else leg for the if check on line 1064?

Reply:
I meant in terms of the logical query semantics. I don't think the extra predicate will impact the cardinality estimate because the value is part of a unique index. I was hesitant to remove the predicate in case it broke something...I'll have a think. I'll either remove it or update the comment to better clarify

Reply:
moved to the else as suggested